### ISSUE TITLE
Not getting the same results between backtest and dryrun

### ISSUE THREAD

**Issue by Chewbye** [freqtrade/freqtrade#7502]
## Describe your environment

  * Operating system: Windows 10 (for backtest), Debian (for dryrun)
  * Python Version: 3.8.13
  * CCXT version: 1.9.84
  * Freqtrade Version: 2022.9.dev-4651a49a4
  
## Describe the problem:
I created a strategy which is running in dryrun mode since a week now.
However there is a difference between backtest performed on the strategy and dryrun results.

### Steps to reproduce:
My strategy: 
[myStrategy.txt](https://github.com/freqtrade/freqtrade/files/9676892/myStrategy.txt)

Config file (I only removed my Telegram tokens from the file): 
[config_dryrun.txt](https://github.com/freqtrade/freqtrade/files/9676871/config_dryrun.txt)

To reproduce:
1) Run the bot in dryrun for some time with command:
freqtrade trade --config config_dryrun.json --strategy myStrategy
2) Backtest the same period of time with 
freqtrade backtesting --timerange 20220923-20220929 --strategy myStrategy --config config_dryrun.json

### Observed Results:

  * What happened?
 Backtesting the strategy I have those results on one week:
![image](https://user-images.githubusercontent.com/5563152/193083606-1b0c93ff-6bbb-41fc-99bc-3e8eb7a653cf.png)
![image](https://user-images.githubusercontent.com/5563152/193084372-5180a730-dbbc-4757-9b1d-ecc67e6c6015.png)

0.56% profit and drawdown 5.55%
Ratio Win/Loss: 610/192

  * What did you expect to happen?
In dryrun mode on one week I get those results:
![image](https://user-images.githubusercontent.com/5563152/193084047-9b43d393-c25d-4aca-9d1b-a447d0d6da6b.png)
 
-8.87% profit and drawdown 10.29%
Ratio Win/Loss: 331/220

I would have expected to have nearly the same results in dryrun than in the backtest but there is a huge difference here.

Is there any explanation why I have this results differences?

Let me know if you need anything else i would have forgot to provide.

Thanks in advance

**Comment by fridary** [comment]
@xmatthias hi, is there any programming or comfortable way to compare live and backtest results? Yes, I can check backtest logs and trade-by-trade via frequi web dashboard, but there are I can not scroll candlestick chart to very past. How do you do that when you need to compare why backtest and live have different profit?

**Comment by xmatthias** [comment]
Backtesting has to take some [assumptions](https://www.freqtrade.io/en/latest/backtesting/#assumptions-made-by-backtesting).

on a quick look, i think a stoploss of 2% is quite tight for a 15m timeframe. 

due to the backtesting assumptions (exit signal before stoploss) - you'll most likely get more regular exits - which will take a different pricing route (priced with open).

---

@fridary i'd ask you to not highjack issues from others in the future. 
It's often dangerous to get the discussion focused on your question, rather than the original one.

in the end, it'll be manual comparison - or via notebooks if you need to look at raw data.

**Comment by Chewbye** [comment]
@xmatthias Thank you for the quick reply.
I will then try to increase the stoploss to 10% and add back an exit signal (initially i had disabled use_exit_signal).
This will be the new settings I will test in dryrun:

```
 minimal_roi = {
        "5760": 0.02,
        "2880": 0.05,
        "1440": 0.10,
        "720": 0.20,
        "0": 0.30
    }

    stoploss = -0.10

use_exit_signal = True
exit_profit_only = False

```

I will run the bot on another week with those parameters and will open a new issue if this is still not working.

**Comment by xmatthias** [comment]
you won't get 100% alignment usually (unless the strategy is very simple) - simply because of the assumptions of backtesting.
With sane values, it is however possible to get very close.

**Comment by Chewbye** [comment]
@xmatthias this strategy seems pretty simple to me.
It is based on only 3 indicators with a fixed stoploss and takeprofit. I don't understand what is not "sane" in this strategy.
After spending hours testing different strategies in backtest and dry run my conclusion is that freqtrade backtesting does not give reliable results.
I will then have to develop my own bot from scratch with a backtest giving realistic results.

**Comment by xmatthias** [comment]
Freqtrade's backtesting is limited by the available data - and is taking the [assumptions as listed in the docs](https://www.freqtrade.io/en/stable/backtesting/#assumptions-made-by-backtesting).

you'll find that all backtesting algos need to take such assumptions (one way or the other) - and all will have a bias one way or the other.
Freqtrade aims to be pessimistic in this - while trying to limit backtestability of certain other functionalities (this is mainly relevant for trailing stoploss features though).

---

Missalignment is usually either user error (maybe Optimized parameters file not transferred, duplicate strategy names, config overwriting stettings in only one instance...) - or also pair selection (if your pairlist is different, the trade sequence will be different, therefore also results will be different).

One additional problem (the most likely in your case) can be unstable indicators (where the result will differ per run). 
We've observed this in the past with MACD (iirc) using a very low parameter (either 0 or 1) - which ta-lib lists as undefined in a code comment (but nothing further) - and therfor returned undefined results (each time slightly different).

I'll have to look this up though - as i'm not sure which parameter it was exactly.

As you say - a simple strategy WILL align if everything is done correctly (we've tested this several times in the past) - however 
why it ain't working for you is a different topic (maybe check the indicator values - on a hunch, i'd distrust both MACD and STOCHF - but i don't ).

> I will then have to develop my own bot from scratch with a backtest giving realistic results.

Talk to you in about a year or maybe two then, on how much effort it was - and how difficult it really is to get alignment between historic and forward-testing.

**Comment by Chewbye** [comment]
@xmatthias 
Thank you the detailed answer.
I will try with different indicators parameters then to see if I have more reliable results.

I will also try to perform a backtesting system on my own during my free time (and I am pretty sure it will not take years to develop 😀) . I will share my repository with you if I succeed :)

**Comment by xmatthias** [comment]
> backtesting system 

well above it was still a bot - which would also involve live trading :) 
Nobody ever said freqtrade's backtesting can't be improved (which might be a better use of your time) - but it's not as simple as it sounds, as you'll for sure see soon.

> I will try with different indicators parameters then to see if I have more reliable results.

in theory, you should see this already by  running the same exact backtest multiple times (make sure to disable cache) - although differences here can be marginal - and therefore quite difficult to spot (if they show at all).

**Comment by xmatthias** [comment]
Found the resource for the macd ta-lib bug [here](https://brookmiles.github.io/freqtrade-stuff/2021/03/29/why-theforce-doesnt-work/).

As your strategy is using exactly the problematic parameter (1 as 3rd parameter) - it's clearly that same bug :)

**Comment by stusiosus** [comment]
to improve the backtest accuracy you can add  for example  --timeframe-detail 5m to the backtest command which uses the smaller candles for better results. See here:https://www.freqtrade.io/en/latest/backtesting/#assumptions-made-by-backtesting
